### DISCUSSION TITLE
Why are there two UUID's in by-component under control-implementation?

### DISCUSSION THREAD

**Discussion by telosBA** [usnistgov/OSCAL#1250] [Q&A] [answered]
`<control-implementation>` `<implemented-requirement>` `<statement>` `<by-component>` currently includes two uuid elements, namely, "component-uuid" and "uuid." Is this redundant? Is there a reason to have both within `<by-component>`?

**Comment by aj-stein-nist** [comment, marked as answer]
One (`uuid`) is the ID of the `by-component` entry itself and one is a reference to the `component` (`component-uuid`)?

Please see the documents on these topics:
- [`component-uuid`](https://pages.nist.gov/OSCAL/reference/1.0.3/system-security-plan/json-reference/#/system-security-plan/control-implementation/implemented-requirements/by-components/component-uuid)
- [`by-component` `uuid`](https://pages.nist.gov/OSCAL/reference/1.0.3/system-security-plan/json-reference/#/system-security-plan/control-implementation/implemented-requirements/by-components/uuid) 

Here is [an example SSP](https://github.com/usnistgov/oscal-content/blob/9dec0d5732415360bf5e794d580c3f8e096be18c/examples/ssp/xml/ssp-example.xml) to help illustrate this example.

**Reply by telosBA** [reply]
What drives the existence of `by-component` `uuid`? Is it the `component-uuid` in combination with the `<statement>` `uuid` the drives the creation of the `by-component` `uuid`? (Important for control inheritance within projects, as `component-uuid` is dynamic.)

**Reply by aj-stein-nist** [reply]
It is a point of identification for the reference for that individual reference to the component when you have n>1 `by-component` usages? In a normal SSP, I perceive there could be dozens or more for a given component. By providing UUIDs you can write application code (if you want, it is not exactly required, but helpful) to target individual references every time you point to a component.

Does that help clarify?

**Reply by GaryGapinski** [reply]
My personal opinion: `by-component` `uuid` should be optional.

**Reply by telosBA** [reply]
> It is a point of identification for the reference for that individual reference to the component when you have n>1 `by-component` usages? In a normal SSP, I perceive there could be dozens or more for a given component. By providing UUIDs you can write application code (if you want, it is not exactly required, but helpful) to target individual references every time you point to a component.
> 
> Does that help clarify?

To confirm, where it is stated that it should not change across versions, if the component changes the statement “Implemented or not implemented”, then the UUID does not change, but if we no longer use that component as a reference, then the UUID is retired. i.e. by-component UUID is unique to the project but the component-uuid is unique to the SITE

**Reply by david-waltermire** [reply]
> My personal opinion: `by-component` `uuid` should be optional.

The `uuid` is there to allow the `by-component` entry to be referenced from the assessment layer. This value is intended to be machine generated. If we make it optional, few implementations will provide it. This will result in an effective denial of service to the assessment layer for most consumers of SSPs. This is why it is required.
